File: Codeforces/Practice/319B. Psychos in a Line.cpp
#include<bits/stdc++.h>

using namespace std;

int n, a[100005], nxt[100005], pre[100005];

int main() {
    scanf("%d", &n);
    for(int i=0; i<n; i++) scanf("%d", &a[i]);
    a[n] = INT_MAX;
    for(int i=0; i<n; i++) nxt[i] = i+1;
    for(int i=1; i<=n; i++) pre[i] = i-1;

    queue<int>q; q.push(-1);
    for(int i=n-1; i>0; i--) if(a[i] < a[ pre[i] ]) q.push(i);

    int ans = 0;
    while(!q.empty()) {
        int in = q.front();
        q.pop();

        if(in == -1) {
            if(q.empty()) break;
            q.push(-1);
            ans++;
            continue;
        }

        nxt[ pre[in] ] = nxt[in];
        pre[ nxt[in] ] = pre[in];
        if(a[ pre[in] ] > a[ nxt[in] ]) q.push(nxt[in]);
    }
    cout << ans << endl;

}

File: Codeforces/Practice/319B. Psychos in a Line.py
from queue import Queue

n = int(input())
a = [int(i) for i in input().split()]
a.append(1000000)

nxt = []
for i in range(n):
    nxt.append(i+1)

pre = [0]
for i in range(1, n+1):
    pre.append(i-1)

# print("BUG")

ans = 0
q = Queue()
q.put(-1);
i = n-1
while i > 0:
    if a[i] < a[i-1]:
        q.put(i)
    i -= 1

while not q.empty():
    cur = q.get()

    # print(cur, pre[cur], nxt[cur])
    # input()

    if cur == -1:
        if(q.empty()):
            break
        q.put(-1)
        ans += 1
        continue

    nxt[ pre[cur] ] = nxt[cur]
    pre[ nxt[cur] ] = pre[cur]
    if(a[ pre[cur] ] > a[ nxt[cur] ]):
        q.put(nxt[cur])

print(ans)
